diff --git a/src/gallium/winsys/virgl/vtest/virgl_vtest_winsys.c b/src/gallium/winsys/virgl/vtest/virgl_vtest_winsys.c
index 2b09608..e207b4a 100644
--- a/src/gallium/winsys/virgl/vtest/virgl_vtest_winsys.c
+++ b/src/gallium/winsys/virgl/vtest/virgl_vtest_winsys.c
@@ -361,7 +361,11 @@ virgl_vtest_winsys_resource_create(struct virgl_winsys *vws,
    }
 
    res->res_handle = handle;
+#ifndef __TERMUX__
    if (map_front_private && res->ptr && res->dt) {
+#else
+   if (map_front_private && res->ptr && res->dt && getenv("TERMUX_ENABLE_GLXCOPYSUBBUFFERMESA_FOR_VTEST")) {
+#endif
       void *dt_map = vtws->sws->displaytarget_map(vtws->sws, res->dt, PIPE_MAP_READ_WRITE);
       uint32_t shm_stride = util_format_get_stride(res->format, res->width);
       util_copy_rect(res->ptr, res->format, shm_stride, 0, 0,
