--- a/meson.build
+++ b/meson.build
@@ -119,7 +119,7 @@
                                    required : get_option('with_wayland') == 'yes',
                                    native : true)
   if wayland_scanner_dep.found()
-    wl_scanner = find_program(wayland_scanner_dep.get_variable(pkgconfig: 'wayland_scanner'))
+    wl_scanner = find_program('@TERMUX_PREFIX@/opt/libwayland/cross/bin/wayland-scanner')
   endif
   WITH_WAYLAND = wayland_dep.found() and wayland_scanner_dep.found()
 endif
