diff --git a/include/android_stub/ndk/sync.h b/include/android_stub/ndk/sync.h
index 38ccb68..5317c10 100644
--- a/include/android_stub/ndk/sync.h
+++ b/include/android_stub/ndk/sync.h
@@ -64,7 +64,7 @@ __BEGIN_DECLS
  *
  * Available since API level 26.
  */
-int32_t sync_merge(const char* name, int32_t fd1, int32_t fd2) __INTRODUCED_IN(26);
+int32_t sync_merge(const char* name, int32_t fd1, int32_t fd2) __INTRODUCED_IN(25);
 
 /**
  * Retrieve detailed information about a sync file and its fences.
@@ -73,7 +73,7 @@ int32_t sync_merge(const char* name, int32_t fd1, int32_t fd2) __INTRODUCED_IN(2
  *
  * Available since API level 26.
  */
-struct sync_file_info* sync_file_info(int32_t fd) __INTRODUCED_IN(26);
+struct sync_file_info* sync_file_info(int32_t fd) __INTRODUCED_IN(25);
 
 /**
  * Get the array of fence infos from the sync file's info.
@@ -97,7 +97,7 @@ static inline struct sync_fence_info* sync_get_fence_info(const struct sync_file
  *
  * Available since API level 26.
  */
-void sync_file_info_free(struct sync_file_info* info) __INTRODUCED_IN(26);
+void sync_file_info_free(struct sync_file_info* info) __INTRODUCED_IN(25);
 
 __END_DECLS
 
